src/sagemaker/image.py

```diff
@@ -0,0 +1,446 @@
+        the Container needs to start.
+        If args.single_model_dir is True, all the hosts will point the opt/ml/model subdir to the first container.
+        That is useful for distributed training, so all the containers can read and write the same checkpoints.
+
+        :param opt_root_folder: root folder with the contents to be mapped to the container
+        :param host: host name of the container
+        :param subdirs: list of subdirs that will be mapped. Example: ['input', 'output', 'model']
+        :return:
+        """
+        volumes = []
+
+        # If it is single mode dir we want to map the same model dir and share between hosts
+        if single_model_dir:
+            host_dir = os.path.join(self.container_root, 'algo-1/model')
+            volume = Volume(host_dir, '/opt/ml/model')
+            volumes.append(volume)
+        else:
+            # else we want to add model to the list of subdirs so it will be created for each container.
+            subdirs.add('model')
+
+        for subdir in subdirs:
+            host_dir = os.path.join(self.container_root, host, subdir)
+            container_dir = '/opt/ml/{}'.format(subdir)
+            volume = Volume(host_dir, container_dir)
+            volumes.append(volume)
+
+        return volumes
+
+
+class _Container(object):
+    def __init__(self, tmpdir, command, startup_delay=5):
+        self.command = command
+        self.compose_file = os.path.join(tmpdir, DOCKER_COMPOSE_FILENAME)
+        self.startup_delay = startup_delay
+        self._process = None
+
+    def up(self):
+        self._process = Popen(self.command)
+        sleep(self.startup_delay)
+
+    def down(self):
+        self._process.terminate()
+        _cleanup()
+
+
+class Volume(object):
+
+    def __init__(self, host_dir, container_dir=None, channel=None):
+        # that is necessary because docker cannot mount mac temp folders.
+        if not container_dir and not channel:
+            raise ValueError('Either container_dir or channel must be declared.')
+
+        if container_dir and channel:
+            raise ValueError('container_dir and channel cannot be declared together.')
+
+        self.container_dir = container_dir if container_dir else os.path.join('/opt/ml/input/data', channel)
+
+        self.host_dir = os.path.join('/private', host_dir) if host_dir.startswith('/var') else host_dir
+
+        self.map = '{}:{}'.format(self.host_dir, self.container_dir)
+
+
+def _cleanup():
+    _chain_docker_cmds('docker images -f dangling=true -q', 'docker rmi -f')
+    _check_output('docker network prune -f'.split(' '))
+
+
+def _chain_docker_cmds(cmd, cmd2):
+    docker_tags = _check_output(cmd).split('\n')
+
+    if any(docker_tags):
+        try:
+            _check_output(cmd2.split() + docker_tags, stderr=STDOUT)
+        except CalledProcessError:
+            pass
+
+
+def _stream_output(cmd):
+    if isinstance(cmd, str):
+        cmd = cmd.split(" ")
```

Comment:
problem here is if there's a double space you'll get empty strings in the list. you probably want just cmd.split()

That being said, this can still blow up. See - https://docs.python.org/2/library/shlex.html#shlex.split